### ISSUE TITLE
Learn vs Infer processor?

### ISSUE THREAD

**Issue by jmrichardson** [microsoft/qlib#602]
Hi,

I am looking at the Alpha158 data handler and it has the following for the Learn and Infer processors:
```
_DEFAULT_LEARN_PROCESSORS = [
    {"class": "DropnaLabel"},
    {"class": "CSZScoreNorm", "kwargs": {"fields_group": "label"}},
]
_DEFAULT_INFER_PROCESSORS = [
    {"class": "ProcessInf", "kwargs": {}},
    {"class": "ZScoreNorm", "kwargs": {}},
    {"class": "Fillna", "kwargs": {}},
]
```

Looking at the code for CSZScoreNorm and ZScoreNorm it appears they do similar normalization (subtract the mean and divide by std along axis 0).  The main difference appears to be that ZScoreNorm has the fit function to "learn" or keep the mean and std  of the fit set (between start and end) which could be used for inference data (new data).  

Why wouldn't ZScoreNorm be used in the _DEFAULT_LEARN_PROCESSORS as well as _DEFAULT_INFER_PROCESSORS? It seems that all the data is being normalized with CSZScoreNorm.  Then at some later point, new data will use the ZScoreNorm which I assume is the same normalization with the exception that it is fitted in the time range?

Also, why wouldn't you also want the Fillna and ProcessInf also in the _DEFAULT_LEARN_PROCESSORS?

Thanks for your help.

**Comment by jmrichardson** [comment]
Ok, I think I understand if someone could please confirm.  The CSZScoreNorm is only for the label which is why it has the the fields_group set to "label".  However, I believe there is no ZScoreNorm for the _DEFAULT_LEARN_PROCESSORS because the default data is already normalized as described here #590?  I am using my own data so I would need ZScoreNorm for both learn and infer if my data is raw?  I am assuming that Inf and NAs have already been fixed in the default data so that is why there is no Fillna and ProcessInf?

**Comment by you-n-g** [comment]
@jmrichardson 
We have to mode for the processors.  The default mode is `PTYPE_A = "append"`.  So the data processors for learning include the processors for inference.

Please refer to the docs here.  Thanks.
https://github.com/microsoft/qlib/blob/main/qlib/data/dataset/handler.py#L430

**Comment by jmrichardson** [comment]
Makes sense now.   Thank you.

**Comment by you-n-g** [comment]
Would you mind sending a PR to improve the docs(or docstrings) about this Question?
Other users who have similar confusion will be grateful to you :)
Thanks

**Comment by jmrichardson** [comment]
Will definitely help with docs soon.

**Comment by jmrichardson** [comment]
Hi @you-n-g ,

I am testing the processors are being executed when running this simple workflow script in the default "append" mode:

```
import qlib
from qlib.contrib.data.handler import Alpha158
from qlib.contrib.model.gbdt import LGBModel
from qlib.utils import exists_qlib_data, init_instance_by_config


data_handler_config = {
    "start_time": "2008-01-01",
    "end_time": "2020-08-01",
    "fit_start_time": "2008-01-01",
    "fit_end_time": "2014-12-31",
    "instruments": 'test',
}

task = {
    "model": {
        "class": "LGBModel",
        "module_path": "qlib.contrib.model.gbdt",
        "kwargs": {
            "loss": "mse",
            "colsample_bytree": 0.8879,
            "learning_rate": 0.0421,
            "subsample": 0.8789,
            "lambda_l1": 205.6999,
            "lambda_l2": 580.9768,
            "max_depth": 8,
            "num_leaves": 210,
            "num_threads": 20,
        },
    },
    "dataset": {
        "class": "DatasetH",
        "module_path": "qlib.data.dataset",
        "kwargs": {
            "handler": {
                "class": "Alpha158",
                "module_path": "qlib.contrib.data.handler",
                "kwargs": data_handler_config,
            },
            "segments": {
                "train": ("2008-01-01", "2014-12-31"),
                "valid": ("2015-01-01", "2016-12-31"),
                "test": ("2017-01-01", "2020-08-01"),
            },
        },
    },
}

if __name__ == "__main__":

    # qlib.init(provider_uri="data/daily/qlib", dataset_cache="SimpleDatasetCache", local_cache_path='data/daily/qlib/cache')
    qlib.init(provider_uri="data/daily/qlib", dataset_cache="SimpleDatasetCache")

    # model initiaiton
    model = init_instance_by_config(task["model"])
    dataset = init_instance_by_config(task["dataset"])

    model.fit(dataset)  # fits on train and valid data segments
    model.predict(dataset)  # predicts test data
```

The default learn and infer processors are:

```
_DEFAULT_LEARN_PROCESSORS = [
    {"class": "DropnaLabel"},
    {"class": "CSZScoreNorm", "kwargs": {"fields_group": "label"}},
]
_DEFAULT_INFER_PROCESSORS = [
    {"class": "ProcessInf", "kwargs": {}},
    {"class": "ZScoreNorm", "kwargs": {}},
    {"class": "Fillna", "kwargs": {}},
]
```

And the Alpha158 has the following set for processors:

```
        infer_processors=[],
        learn_processors=_DEFAULT_LEARN_PROCESSORS,
```

According to process_type = "append" mode (which i have verified is append), both the infer and learn processors should be executed on the data.  However, only the learn processors appear to be executed.  Here is the output of the script:

```
[13404:MainThread](2021-09-16 13:19:26,155) INFO - qlib.Initialization - [config.py:350] - default_conf: client.
[13404:MainThread](2021-09-16 13:19:28,203) WARNING - qlib.Initialization - [config.py:375] - redis connection failed(host=127.0.0.1 port=6379), DiskExpressionCache will not be used!
[13404:MainThread](2021-09-16 13:19:28,216) INFO - qlib.SimpleDatasetCache - [cache.py:1074] - DatasetCache directory: C:\Users\john\.cache\qlib_simple_cache, modify the cache directory via the local_cache_path in the config
[13404:MainThread](2021-09-16 13:19:28,219) INFO - qlib.Initialization - [__init__.py:58] - qlib successfully initialized based on client settings.
[13404:MainThread](2021-09-16 13:19:28,221) INFO - qlib.Initialization - [__init__.py:60] - data_path={'__DEFAULT_FREQ': WindowsPath('D:/Projects/aim/data/daily/qlib')}
D:\Anaconda3\envs\aim\lib\site-packages\pyqlib-0.7.1-py3.8-win-amd64.egg\qlib\utils\__init__.py:745: FutureWarning: MultiIndex.is_lexsorted is deprecated as a public function, users should use MultiIndex.is_monotonic_increasing instead.
  if idx.is_monotonic_increasing and not (isinstance(idx, pd.MultiIndex) and not idx.is_lexsorted()):
[13404:MainThread](2021-09-16 13:19:28,268) INFO - qlib.timer - [log.py:113] - Time cost: 0.042s | Loading data Done
[13404:MainThread](2021-09-16 13:19:28,278) INFO - qlib.timer - [log.py:113] - Time cost: 0.006s | DropnaLabel Done
[13404:MainThread](2021-09-16 13:19:43,768) INFO - qlib.timer - [log.py:113] - Time cost: 15.490s | CSZScoreNorm Done
[13404:MainThread](2021-09-16 13:19:43,769) INFO - qlib.timer - [log.py:113] - Time cost: 15.501s | fit & process data Done
[13404:MainThread](2021-09-16 13:19:43,769) INFO - qlib.timer - [log.py:113] - Time cost: 15.544s | Init data Done
```

Notice only Dropnalabel and CSXScoreNorm and not the inference processors.  I also put a debug breakpoint on the code for the inference processors and they are never executed.

If I change the infer_processors for Alpha168 to _DEFAULT_INFER_PROCESSORS:

```
class Alpha158(DataHandlerLP):
    def __init__(
        self,
        instruments="csi500",
        start_time=None,
        end_time=None,
        freq="day",
        infer_processors=[],
        learn_processors=_DEFAULT_LEARN_PROCESSORS + _DEFAULT_INFER_PROCESSORS,

```

Output of script (Note I commented out ProcessInf because it takes a very long time):

```
[18468:MainThread](2021-09-16 13:37:39,271) INFO - qlib.Initialization - [config.py:350] - default_conf: client.
[18468:MainThread](2021-09-16 13:37:41,321) WARNING - qlib.Initialization - [config.py:375] - redis connection failed(host=127.0.0.1 port=6379), DiskExpressionCache will not be used!
[18468:MainThread](2021-09-16 13:37:41,337) INFO - qlib.SimpleDatasetCache - [cache.py:1074] - DatasetCache directory: C:\Users\john\.cache\qlib_simple_cache, modify the cache directory via the local_cache_path in the config
[18468:MainThread](2021-09-16 13:37:41,340) INFO - qlib.Initialization - [__init__.py:58] - qlib successfully initialized based on client settings.
[18468:MainThread](2021-09-16 13:37:41,341) INFO - qlib.Initialization - [__init__.py:60] - data_path={'__DEFAULT_FREQ': WindowsPath('D:/Projects/aim/data/daily/qlib')}
D:\Anaconda3\envs\aim\lib\site-packages\pyqlib-0.7.1-py3.8-win-amd64.egg\qlib\utils\__init__.py:745: FutureWarning: MultiIndex.is_lexsorted is deprecated as a public function, users should use MultiIndex.is_monotonic_increasing instead.
  if idx.is_monotonic_increasing and not (isinstance(idx, pd.MultiIndex) and not idx.is_lexsorted()):
[18468:MainThread](2021-09-16 13:37:41,380) INFO - qlib.timer - [log.py:113] - Time cost: 0.034s | Loading data Done
[18468:MainThread](2021-09-16 13:37:41,389) INFO - qlib.timer - [log.py:113] - Time cost: 0.006s | DropnaLabel Done
[18468:MainThread](2021-09-16 13:37:53,009) INFO - qlib.timer - [log.py:113] - Time cost: 11.619s | CSZScoreNorm Done
D:\Anaconda3\envs\aim\lib\site-packages\pyqlib-0.7.1-py3.8-win-amd64.egg\qlib\data\dataset\processor.py:236: RuntimeWarning: Mean of empty slice
  self.mean_train = np.nanmean(df[cols].values, axis=0)
C:\Users\john\AppData\Roaming\Python\Python38\site-packages\numpy\lib\nanfunctions.py:1664: RuntimeWarning: Degrees of freedom <= 0 for slice.
  var = nanvar(a, axis=axis, dtype=dtype, out=out, ddof=ddof,
[18468:MainThread](2021-09-16 13:38:08,708) INFO - qlib.timer - [log.py:113] - Time cost: 15.698s | ZScoreNorm Done
[18468:MainThread](2021-09-16 13:38:12,870) INFO - qlib.timer - [log.py:113] - Time cost: 4.160s | Fillna Done
[18468:MainThread](2021-09-16 13:38:12,872) INFO - qlib.timer - [log.py:113] - Time cost: 31.490s | fit & process data Done
[18468:MainThread](2021-09-16 13:38:12,873) INFO - qlib.timer - [log.py:113] - Time cost: 31.527s | Init data Done
```

Am I doing something wrong or should I just include the processors needed for each learn and infer?

**Comment by you-n-g** [comment]
@jmrichardson 
Please refer to the docs here https://github.com/microsoft/qlib/blob/main/qlib/data/dataset/handler.py#L311
> infer_processors=[],
I noticed that you didn't pass any processor to these parameters.

So `shared_processors + infer_processors + learn_processors` will be `'[] + [] + _DEFAULT_LEARN_PROCESSORS`
The results are expected.
